Yoast/wordpress-seo
# Lazily update indexable hierarchy

## Context


* When we change the taxonomy to show for a certain post type we want to remove the indexable hierarchy records for the belonging indexables. This allows us to recalculate the hierarchy for these posts

## Summary


This PR can be summarized in the following changelog entry:

* N/A - Fixes a bug where wrong breadcrumbs are shown when changing the primary taxonomy to show for the breadcrumbs.

## Relevant technical choices:

* Decided to add extra tests for the hierarchy repository, because I had the context.
* Temporary added `codeCoverageIgnore` for the query method, because testing this one was too complex. In mockery aliasing, overloading doesn't work. Also in combination with `@runInSeperateProcess` won't work. This needs more investigation... (but that is for a later moment)
* The `-1` check for ancestor_id has been changed to `0` because the ancester_id column is unsigned which means that negative integers aren't allowed. 

## Test instructions

This PR can be tested by following these steps:

* Save a post with multiple categories attached and one selected as the primary.
* Open your database and check which indexable belongs to the post.
* Open your database and view the table `{some_prefix}_indexable_hierarchy` and see some records for the indexable being present.
* Go to the breadcrumbs settings (Yoast SEO -> Search Appearance -> Breadcrumbs and change the `Taxonomy to show in breadcrumbs for content types` for the post from category to tag.
* In the database hierarchy table the records should be gone.
* Visit the post and make sure new records are made again.
* In the source of the post (browser) the breadcrumbs structure should be the right one.

## UI changes

## Documentation

## Quality assurance


Fixes #14558
